migrate web-client development from grunt to vite

return42 opened the pull request:
## What does this PR do?

Migrate web client development (simple theme) from grunt to [Vite](https://vite.dev/).

Vite consists of a builder and a server, this PR only implements the build, what grunt has done so far. We will integrate the vite server into our development processes in a followup PR.

## Why is this change important?

1. grunt and its plugins are no longer maintained  
2. we need a modern developer environment for front-end (FE) development

## How to test this PR locally?

Start a server (`make run`) and see if there are no changes in the frontend.  Apart from a few small bugs that have been fixed around the icons, this PR should not bring any changes from the user's point of view.

To test FE development jump to folder `client/simple` and have a look at the README there.

## Notes

The switch to vite in this PR will enable us to develop our FE faster and more securely in future (e.g. we do no longer rely on unmaintained components and can benefit from state of the art technologies).

This PR deliberately avoids changes to the JS code and adopts it 1:1 for the time being. In further PRs we should then restructure the JS code base to JS modules (and imports/exports).

The migration to vite currently still has one catch: I have not been able to generate the source maps for the CSS that are generated from the LESS ... perhaps we can find a solution for this in the course of the review ..

- https://github.com/vitejs/vite/discussions/13845#discussioncomment-11992084

## Related issues

- https://github.com/searxng/searxng/issues/559
- https://github.com/searxng/searxng/issues/141

-----

Closes: #141

return42 committed:
[web-client] simple theme: build & development env based on vite.

Vite is a build tool, it consists of two major parts [1]:

- A dev server that provides rich feature enhancements over native ES modules,
  for example extremely fast Hot Module Replacement (HMR).

- A build command that bundles your code with Rollup, pre-configured to output
  highly optimized static assets for production.

[1] https://vite.dev/guide/

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[web-client] simple theme: move sources to client/simple/src

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[web-client] simple theme: vite build -- CSS sourcemaps still missing!!!

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[web-client] simple theme: eslint JS source files

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[web-client] simple theme: move src/brand to client/simple/src

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[web-client] simple theme: vite plugins plg_svg2png & plg_svg2svg

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[web-client] simple theme: refactor jinja_svg_catalog and cleanup icon names

This patch implements the template generation of the template:

  searx/templates/simple/icons.html

by the way the icon set (the icon names) has been normalized:

  film-outline   --> film
  magnet-outline --> magnet
  ..
  warning        --> alert

Some missing (categorie) icons had been added.

Some of the ionicons are not suitable for a dark theme, we fixed the svg
manually in src/svg/ionicons:

- https://github.com/searxng/searxng/pull/4284#issuecomment-2680550342

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[web-client] simple theme: stylelint LESS source files

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[web-client] simple theme: stylelint-prettier LESS source files

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[web-client] simple theme: move SVGs to client/simple/src and deploy to static by plg_svg2svg

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[web-client] simple theme: final build by vite & clean up grunt leftovers

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[web-client] integrate the migrated theme in the development workflows

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[web-client] integrate the migrated theme in the github CI

- upgrade to ubuntu-24.04 to get NodeJS v20
- remove DEBUG (V=1) environment / in a ViteJS env the DEBUG environment is
  reserved for interactive debugging tasks (not for verbose build messages).

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

Bnyro reviewed (commented):
It looks like the JavaScript on the results page is currently not being loaded properly - I can't tell why exactly yet.

To reproduce: Try to `Shift`-select additional categories with the left mouse button - works on the latest master but not with this PR.

Bnyro commented on client/simple/package.json:
```diff
@@ -0,0 +1,40 @@
+  "scripts": {
+    "clean": "rm -Rf node_modules",
+    "build": "node theme_icons.js && vite build",
+    "fix": "eslint --fix && stylelint --fix strict 'src/**/*.{css,scss,sass,less,styl,vue,svelte}'",
```
`sass,scss,styl,vue,svelte` are probably not needed since we don't use these anywhere? 

We only use `less` (and maybe also `css`).

return42 replied:
By default stylelint checks all files matching js glob pattern in `./`

    ./**/*.{css,scss,sass,less,styl,vue,svelte}

I left it at default, with the only difference that we have our sources in `/src`

    src/**/*.{css,scss,sass,less,styl,vue,svelte}

If you insist, we change it, but in my opinion, the advantage is that if a PR adds new file types in the future, they will be checked automatically (without any changes to the command here) by stylelint ... and there is no disadvantage.

Bnyro commented on client/simple/package.json:
```diff
@@ -0,0 +1,40 @@
+    "webpack-cli": "^6.0.1"
+  },
+  "dependencies": {
+    "autocomplete-js": "^2.7.1"
```
This dependency is unmaintained since 6 years, we shouldn't use this anymore (not related to this PR) ...
(source: https://www.npmjs.com/package/autocomplete-js)

I'll make a PR to replace it in some time...

return42 replied:
Thats true .. we have such old unmaintained libs .. 

> I'll make a PR to replace it in some time...

:rocket: .. and we should take in mind there are open tasks --> https://github.com/searxng/searxng/issues?q=label%3A%22area%3A%20autocomplete%22%20state%3Aopen

But first lets migrate the code "as is" into the new environment for our FE development (vite), than we can migrate the JS code base to JS modules .. and when we have JS modules we can implement a new autocompleter and improve other client stuff.

Bnyro replied:
That warning was too late, already implemented in https://github.com/searxng/searxng/pull/4318.

I can rebase it after this PR is merged though.

return42 replied:
> I can rebase it after this PR is merged though.

I will see what will be easier .. but not today :-)

return42 committed:
[fix] simple template - video.html fix button icon to icon_small

BTW: Rename the manufacturer-specific name ``ion-icon`` of the icon set in
SerXNG to ``sxng-icon-set`` --> The icon set can be assembled together as
desired in the FE and should therefore not have a manufacturer-specific name in
the style and template definitions.

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[upd] npm: Bump eslint from 9.19.0 to 9.20.1 in /client/simple

Related:

- https://github.com/searxng/searxng/pull/4322

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[build] /static

return42 commented on client/simple/theme_icons.js:
```diff
@@ -0,0 +1,84 @@
+
+
+/** @type import("./tools/jinja_svg_catalog.js").IconSet */
+const simple_icons = [
+  {
+    base: resolve(HERE, "node_modules/ionicons/dist/svg"),
+    set: {
+      "alert": "alert-outline.svg",
+      "appstore": "apps-outline.svg",
+      "book": "book-outline.svg",
+      "close": "close-outline.svg",
+      "download": "download-outline.svg",
+      "ellipsis-vertical": "ellipsis-vertical-outline.svg",
+      "file-tray-full": "file-tray-full-outline.svg",
+      "film": "film-outline.svg",
+      "globe": "globe-outline.svg",
+      "heart": "heart-outline.svg",
+      "image": "image-outline.svg",
+      "layers": "layers-outline.svg",
+      "leecher": "arrow-down.svg",
+      "location": "location-outline.svg",
+      "magnet": "magnet-outline.svg",
+      "musical-notes": "musical-notes-outline.svg",
+      "navigate-down": "chevron-down-outline.svg",
+      "navigate-left": "chevron-back-outline.svg",
+      "navigate-right": "chevron-forward-outline.svg",
+      "navigate-up": "chevron-up-outline.svg",
+      "people": "people-outline.svg",
+      "play": "play-outline.svg",
+      "radio": "radio-outline.svg",
+      "save": "save-outline.svg",
+      "school": "school-outline.svg",
+      "search": "search-outline.svg",
+      "seeder": "swap-vertical.svg",
+      "settings": "settings-outline.svg",
+      "tv": "tv-outline.svg",
+    },
```
@vojkovic: while we are just talking about icons .. this PR also fix some very old issues with some missing SVGs: e.g. SVGs for `seeder`, `lecher` and *file size* (`save`) had been missed in the past ..


![grafik](https://github.com/user-attachments/assets/dd9df0d0-c98e-4a7d-905d-1a2a13078385)


![grafik](https://github.com/user-attachments/assets/cb14ba57-949f-4e27-80f7-4f9166640e9d)

vojkovic replied:
Very nice - great work with this mate.

return42 commented on utils/lib_sxng_themes.sh:
```diff
@@ -1,100 +1,36 @@
-    dump_return $?
-}
-
-themes.live() {
-    local LIVE_THEME="${LIVE_THEME:-${1}}"
-    case "${LIVE_THEME}" in
-        simple)
-            theme="searx/static/themes/${LIVE_THEME}"
-            ;;
```
> I read over the github discussion and even after checking the source of vite, I can't get them to build either.

@vojkovic many thanks for your research on this .. I guess the issue is related to the *minification* of the CSS .. but for the **build phase** I prefer to to *minify* ..

> Vite consists of a builder and a server, this PR only implements the build,

in a follow up PR I will integrate vite's development server in a (new) `themes.live` command .. then, in the **development phase** we don't need to *minify* and have CSS source maps back .. at least I hope so ;-)

----

BTW: we already have a similar issue with LESS & CSS

- https://github.com/searxng/searxng/pull/4215
- https://github.com/less/less-plugin-clean-css/issues/42#issuecomment-2602524014

The problem may lie in a [sub-sub-sub dependency](https://images.ctfassets.net/vkoe68wv76dt/2zWABwW9gb33FGGWBflZ3y/143994d9368dd33f1dbe31aba17ec517/node-modules-app-performance_.png) and therefore affects both (vite and less-plugin-clean-css)... but I can only speculate about that ..

Bnyro reviewed (approved):
Tested again, looks good to me now! :+1:
